File: c++/17. 电话号码的字母组合/17. 电话号码的字母组合.cpp
class Solution {
public:
    vector<string> letterCombinations(string digits) {
        map<int, string> m;
        m[2] = "abc";
        m[3] = "def";
        m[4] = "ghi";
        m[5] = "jkl";
        m[6] = "mno";
        m[7] = "pqrs";
        m[8] = "tuv";
        m[9] = "wxyz";
        vector<string> ans;
        if(digits == "") return ans;
        int num = digits[0] - '0';
        for(int i = 0; i < m[num].size(); i++){
            string tmp;
            tmp.push_back(m[num][i]);
            ans.push_back(tmp);
            
        }
        for(int i = 1; i < digits.size(); i++){
            num = digits[i] - '0';
            vector<string> tmp = ans;
            for(int j = 0; j < ans.size(); j++){
                ans[j].push_back(m[num][0]);
            }
            int n = ans.size();
            for(int k = 1; k < m[num].size(); k++){
                for(int j = 0; j < n; j++){
                    string elem = tmp[j] + m[num][k];
                    ans.push_back(elem);
                }
            }
        }
        return ans;

    }
};

File: Java/17_电话号码的字母组合/17. 电话号码的字母组合.java
class Solution {
    Map<String, String> m = new HashMap<>();
    List<String> ans = new ArrayList<>();
    void dfs(String digits, int idx, String tmp, Map<String, String> m, int n){
        if(n == digits.length()){
           
            ans.add(tmp);
            return;
        }
       
            String num = digits.substring(idx, idx + 1);
            
            String str = m.get(num);
            
            for(int j = 0; j < str.length(); j++){
                
                tmp += (str.charAt(j));
                dfs(digits, idx + 1, tmp, m, n + 1);
                tmp = tmp.substring(0, tmp.length() - 1);
            }
       
    }
    public List<String> letterCombinations(String digits) {
        if(digits.length() == 0) return ans;
        m.put("2", "abc");
        m.put("3", "def");
        m.put("4", "ghi");
        m.put("5", "jkl");
        m.put("6", "mno");
        m.put("7", "pqrs");
        m.put("8", "tuv");
        m.put("9", "wxyz");
        String tmp = "";
        
        dfs(digits, 0, tmp, m, 0);
        return ans;
    }
}